File: pkg/client/loader/token.go
package loader

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strings"
)

type tokenConf struct {
	serverURL            string
	grpcBroadcastAddress string
	tenantId             string
}

func getConfFromJWT(token string) (*tokenConf, error) {
	claims, err := extractClaimsFromJWT(token)
	if err != nil {
		return nil, err
	}

	serverURL, ok := claims["server_url"].(string)
	if !ok {
		return nil, fmt.Errorf("server_url claim not found")
	}

	grpcBroadcastAddress, ok := claims["grpc_broadcast_address"].(string)
	if !ok {
		return nil, fmt.Errorf("grpc_broadcast_address claim not found")
	}

	tenantId, ok := claims["sub"].(string)

	if !ok {
		return nil, fmt.Errorf("sub claim not found")
	}

	return &tokenConf{
		serverURL:            serverURL,
		grpcBroadcastAddress: grpcBroadcastAddress,
		tenantId:             tenantId,
	}, nil
}

func extractClaimsFromJWT(token string) (map[string]interface{}, error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, fmt.Errorf("invalid token format")
	}

	claimsData, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, err
	}

	var claims map[string]interface{}
	err = json.Unmarshal(claimsData, &claims)
	if err != nil {
		return nil, err
	}

	return claims, nil
}

File: sdks/python/hatchet_sdk/token.py
import base64

from pydantic import BaseModel


class Claims(BaseModel):
    sub: str
    server_url: str
    grpc_broadcast_address: str


def get_tenant_id_from_jwt(token: str) -> str:
    return extract_claims_from_jwt(token).sub


def get_addresses_from_jwt(token: str) -> tuple[str, str]:
    claims = extract_claims_from_jwt(token)

    return claims.server_url, claims.grpc_broadcast_address


def extract_claims_from_jwt(token: str) -> Claims:
    parts = token.split(".")
    if len(parts) != 3:
        raise ValueError("Invalid token format")

    claims_part = parts[1]
    claims_part += "=" * ((4 - len(claims_part) % 4) % 4)  # Padding for base64 decoding
    claims_data = base64.urlsafe_b64decode(claims_part)

    return Claims.model_validate_json(claims_data)
